JCRVLT-789: AggregateImpl - avoid iterating over sibling nodes

reschke opened the pull request:
Work in progress - this passes both existing tests and the new test; feedback appreciated.

TODO: Javadoc, logging cleanup , more exception handling, maybe rename new methods.

kwin commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/WorkspaceFilter.java:
```diff
@@ -102,6 +103,19 @@ public interface WorkspaceFilter extends Dumpable {
+     * @return first path segments of non-matched paths, or {@code null} when result set
+     * cam not be computed.
+     */
+    default @Nullable Set<String> getFirstChildNamesOfRootsBelowPath(@NotNull String path) {
```
maybe `getDirectChildNamesTowardFilterRoots`?

kwin commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/FilterSet.java:
```diff
@@ -239,12 +243,45 @@ public boolean covers(@NotNull String path) {
+     * @param path Path to check
+     * @return first path segment of non-matched path, or {@code null} when path not ancestor
+     */
+    public @Nullable String getChildNameBelowFilterRoot(@NotNull String path) {
```
likewise `getDirectChildNameTowardsFilterRoot`

kwin commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/WorkspaceFilter.java:
```diff
@@ -102,6 +103,19 @@ public interface WorkspaceFilter extends Dumpable {
+     * to the result set.
+     *
+     * @param path Path to check
+     * @return first path segments of non-matched paths, or {@code null} when result set
```
The condition when null is returned is not really clear. What if only some roots are no offsprings of the given path?

reschke replied:
It would return null, because in that case we can't optimize? So clarify the docs?

reschke replied:
Attempted to clariffy, please check.

(Note that there may be room for improvements here, but  for now the change is sufficient for my use case)

kwin replied:
For me there is three conditions:

1. all filter roots are below given path
2. a filter root is not related (i.e. neither ancestor nor descendent)
3. a filter root is an ancestor of this path

Are you sure that both 2. and 3. should lead to the same return value `null`?

reschke replied:
Well - `null` is the safe approach, because it means doing exactly what we did before.

And no, I'm not sure. I'm trying to improve one specific use case (for which I happen to have an integration test to verify). We probably can improve more cases, but my personal preference would be to do that when we figure out that it's needed - for instance when the current improvement can't be applied because there is a second completely unrelated filter.

Does this make sense?

mbaedke commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/FilterSet.java:
```diff
@@ -239,12 +243,45 @@ public boolean covers(@NotNull String path) {
 
     /**
-     * Checks if the given item is an ancestor of the root node.
+     * Checks if the given item is an ancestor of the filter's root path.
```
s/given item/given path

reschke replied:
it's consistent with the method above; should I change both?

kwin replied:
I think what was meant was rather 
`Checks if the given path is an ancestor of the filter's root path`. A (string) path is given, not a JCR item!

reschke replied:
Yes. What I meant was: this is consistent with the Javadoc for other methods here. Should I use the opportunity to fix those as well?

kwin commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/FilterSet.java:
```diff
@@ -239,12 +243,45 @@ public boolean covers(@NotNull String path) {
+            result = rel.isEmpty() ? null : rel;
+        }
+
+        log.debug("getChildNameBelowRoot(root={}, path={}) -> {}", rootMatch, pathMatch, result);
```
This is inconsistent with the current name of the method.

reschke replied:
ack

kwin commented on vault-core-it/vault-core-integration-tests/src/main/java/org/apache/jackrabbit/vault/packaging/integration/ManySiblingsIT.java:
```diff
@@ -0,0 +1,104 @@
+
+    // JCRVLT-789
+    @Test
+    public void testManySiblings() throws RepositoryException, IOException {
```
I think some more tests with filters above or outside the path with the many children should be added (or this one being extended).

reschke replied:
I have now added a test for case 2, as expected the outcome is unnessary iteration.

reschke commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/FilterSet.java:
```diff
@@ -239,12 +243,45 @@ public boolean covers(@NotNull String path) {
+        String rootMatch = appendSlashIfNeeded(root);
+        String pathMatch = appendSlashIfNeeded(path);
+
+        if (rootMatch.startsWith(pathMatch)) {
```
Well, we can avoid those for the comparison (which internally normalizes), but we'd need it anyway for the string operations below.

It would be nice to have something like "getDescendantRelativeTo", but right now it's not there (could be inspired by nio Path https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#relativize-java.nio.file.Path-.

reschke commented:
If have now refactored tests, added coverage for more combinations, and added a case "we know no children are covered". I *believe* the "null" case is gone now, but would like to get feedback before I kill it.

@cc @kwin and @mbaedke

kwin commented on vault-core-it/vault-core-integration-tests/src/main/java/org/apache/jackrabbit/vault/packaging/integration/SiblingIterationIT.java:
```diff
@@ -0,0 +1,149 @@
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests checking whether siblings
```
this sentence should have a verb :-)

reschke replied:
Indeed.

kwin commented on vault-core/src/main/java/org/apache/jackrabbit/vault/fs/api/FilterSet.java:
```diff
@@ -239,12 +243,68 @@ public boolean covers(@NotNull String path) {
+     * @param path Path to check
+     * @return first path segment of non-matched path, or {@code null} when path not ancestor
+     */
+    public @Nullable String getDirectChildNameTowardsFilterRoot(@NotNull String path) {
```
A Unit test would be nice for this.

reschke replied:
In the unit tests, I assume?

reschke replied:
See https://github.com/apache/jackrabbit-filevault/pull/362/commits/788f046f872005736806c458a43954edc18cd597

kwin commented:
Looks good now, but ASF Jenkins crashes when executing ITs. Now sure it is related to the one you added.

reschke commented:
master seemt to have a similar issue: https://ci-builds.apache.org/job/Jackrabbit/job/filevault/job/master/369/cloudbees-pipeline-explorer/

kwin commented:
Hopefully fixed by https://github.com/apache/jackrabbit-filevault/commit/a5a62070cc93151bf12d08d579bad2cf4c474b00.
